src/System.IO.FileSystem/src/System/IO/PathHelpers.cs

```diff
@@ -38,5 +41,146 @@ internal static bool EndsInDirectorySeparator(string path)
+
+            if (string.IsNullOrEmpty(second))
+                return first;
+
+            return CombineNoChecksInternal(first.AsReadOnlySpan(), second.AsReadOnlySpan());
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe static string CombineNoChecksInternal(ReadOnlySpan<char> first, ReadOnlySpan<char> second)
+        {
+            Debug.Assert(first.Length > 0 && second.Length > 0, "should have dealt with empty paths");
+
+            bool hasSeparator = first[first.Length - 1] == Path.DirectorySeparatorChar
+                || second[0] == Path.DirectorySeparatorChar;
+
+            int totalLength = first.Length + second.Length + (hasSeparator ? 0 : 1);
+            string fullPath = new string('\0', totalLength);
+            fixed (char* f = fullPath)
+            {
+                Span<char> pathSpan = new Span<char>(f, totalLength);
+                first.CopyTo(pathSpan);
+                if (!hasSeparator)
+                    pathSpan[first.Length] = Path.DirectorySeparatorChar;
+                second.CopyTo(pathSpan.Slice(first.Length + (hasSeparator ? 0 : 1)));
+            }
+            return fullPath;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe static string CombineNoChecksInternal(ReadOnlySpan<char> first, ReadOnlySpan<char> second, ReadOnlySpan<char> third)
+        {
+            Debug.Assert(first.Length > 0 && second.Length > 0 && third.Length > 0, "should have dealt with empty paths");
+
+            bool firstHasSeparator = first[first.Length - 1] == Path.DirectorySeparatorChar
+                || second[0] == Path.DirectorySeparatorChar;
+            bool thirdHasSeparator = second[second.Length - 1] == Path.DirectorySeparatorChar
+                || third[0] == Path.DirectorySeparatorChar;
+
+            int totalLength = first.Length + second.Length + third.Length + (firstHasSeparator ? 0 : 1) + (thirdHasSeparator ? 0 : 1);
+            string fullPath = new string('\0', totalLength);
+            fixed (char* f = fullPath)
+            {
+                Span<char> pathSpan = new Span<char>(f, totalLength);
+                first.CopyTo(pathSpan);
+                if (!firstHasSeparator)
+                    pathSpan[first.Length] = Path.DirectorySeparatorChar;
+                second.CopyTo(pathSpan.Slice(first.Length + (firstHasSeparator ? 0 : 1)));
+                if (!thirdHasSeparator)
+                    pathSpan[pathSpan.Length - third.Length - 1] = Path.DirectorySeparatorChar;
+                third.CopyTo(pathSpan.Slice(pathSpan.Length - third.Length));
+            }
+            return fullPath;
+        }
+
+        /// <summary>
+        /// Returns true if the file name is "." or ".."
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe bool IsDotOrDotDot(ReadOnlySpan<char> fileName)
+        {
+            return !(fileName.Length > 2
```

Comment:
isn't this more efficient (unless jitter unravels by de morgans law?)
```c#
            return fileName.Length <3 && fileName[0] != '.' && (fileName.Length != 2 || fileName[1] == '.'));
```

Reply:
Don't know- I'll play with it

Reply:
Well it's no worse. You get my point, that wrapping an expression in ! means it could potentially not shortcircuit.